File: src/main/java/de/moritzf/quota/idea/QuotaSettingsConfigurable.java
package de.moritzf.quota.idea;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.ui.components.JBTextArea;
import com.intellij.ui.components.JBTextField;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.messages.MessageBusConnection;
import org.jetbrains.annotations.Nullable;

import javax.swing.JButton;
import javax.swing.JComponent;
import javax.swing.JComboBox;
import javax.swing.JPanel;
import java.awt.FlowLayout;

/**
 * Settings UI that manages authentication actions and shows latest quota payload data.
 */
public final class QuotaSettingsConfigurable implements Configurable {
    private JPanel panel;
    private JBTextField accountIdField;
    private JBTextField emailField;
    private JBTextArea responseArea;
    private JBLabel loginHeaderLabel;
    private JBLabel statusLabel;
    private JComboBox<QuotaDisplayMode> displayModeComboBox;
    private JButton loginButton;
    private JButton cancelLoginButton;
    private JButton logoutButton;
    private MessageBusConnection connection;

    @Override
    public String getDisplayName() {
        return "OpenAI Usage Quota";
    }

    @Override
    public @Nullable JComponent createComponent() {
        accountIdField = new JBTextField();
        accountIdField.setEditable(false);
        emailField = new JBTextField();
        emailField.setEditable(false);
        responseArea = new JBTextArea();
        responseArea.setEditable(false);
        responseArea.setLineWrap(false);
        loginHeaderLabel = new JBLabel();
        statusLabel = new JBLabel();
        displayModeComboBox = new ComboBox<>(QuotaDisplayMode.values());
        loginButton = new JButton("Log In");
        cancelLoginButton = new JButton("Cancel Login");
        logoutButton = new JButton("Log Out");

        loginButton.addActionListener(event -> {
            QuotaAuthService authService = QuotaAuthService.getInstance();
            if (authService.isLoggedIn()) {
                updateAuthUi();
                return;
            }
            loginButton.setEnabled(false);
            statusLabel.setText("Opening browser...");
            authService.startLoginFlow(result -> {
                if (panel == null) {
                    return;
                }
                ApplicationManager.getApplication().invokeLater(() -> {
                    if (panel == null || statusLabel == null || loginButton == null || logoutButton == null) {
                        return;
                    }
                    if (result.success) {
                        statusLabel.setText("Logged in");
                        QuotaUsageService.getInstance().refreshNowAsync();
                    } else {
                        String message = result.message != null ? result.message : "Login failed";
                        statusLabel.setText("Login failed");
                        Messages.showErrorDialog(panel, message, "OpenAI Login");
                    }
                    loginButton.setEnabled(true);
                    updateAuthUi();
                    updateAccountFields();
                }, ModalityState.stateForComponent(panel));
            });
            updateAuthUi();
        });

        cancelLoginButton.addActionListener(event -> {
            boolean aborted = QuotaAuthService.getInstance().abortLogin("Login canceled");
            statusLabel.setText(aborted ? "Login canceled" : "No login in progress");
            updateAuthUi();
        });

        logoutButton.addActionListener(event -> {
            QuotaAuthService.getInstance().clearCredentials();
            updateAuthUi();
            updateAccountFields();
            QuotaUsageService.getInstance().refreshNowAsync();
        });

        JPanel authPanel = new JBPanel<>(new FlowLayout(FlowLayout.LEFT, 8, 0));
        authPanel.add(loginButton);
        authPanel.add(cancelLoginButton);
        authPanel.add(logoutButton);

        JBScrollPane responseScroll = new JBScrollPane(responseArea);
        responseScroll.setHorizontalScrollBarPolicy(JBScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        responseScroll.setVerticalScrollBarPolicy(JBScrollPane.VERTICAL_SCROLLBAR_AS_NEEDED);

        panel = FormBuilder.createFormBuilder()
                .addLabeledComponent("Status bar display:", displayModeComboBox)
                .addSeparator()
                .addComponent(loginHeaderLabel)
                .addComponent(authPanel)
                .addSeparator()
                .addLabeledComponent("Account ID:", accountIdField)
                .addLabeledComponent("Email:", emailField)
                .addSeparator()
                .addLabeledComponent("Last quota response (JSON):", responseScroll)
                .getPanel();

        connection = ApplicationManager.getApplication().getMessageBus().connect();
        connection.subscribe(QuotaUsageListener.TOPIC, (QuotaUsageListener) (quota, error) -> {
            if (panel == null) {
                return;
            }
            ApplicationManager.getApplication().invokeLater(() -> {
                if (panel == null || responseArea == null || accountIdField == null || emailField == null) {
                    return;
                }
                updateAccountFields();
                updateResponseArea();
            }, ModalityState.stateForComponent(panel));
        });

        reset();
        return panel;
    }

    @Override
    public boolean isModified() {
        if (displayModeComboBox == null) {
            return false;
        }
        QuotaDisplayMode selected = (QuotaDisplayMode) displayModeComboBox.getSelectedItem();
        QuotaDisplayMode saved = QuotaSettingsState.getInstance().getStatusBarDisplayMode();
        return selected != saved;
    }

    @Override
    public void apply() {
        if (displayModeComboBox == null) {
            return;
        }
        QuotaDisplayMode selected = (QuotaDisplayMode) displayModeComboBox.getSelectedItem();
        if (selected == null) {
            return;
        }
        QuotaSettingsState state = QuotaSettingsState.getInstance();
        QuotaDisplayMode current = state.getStatusBarDisplayMode();
        if (selected != current) {
            state.setStatusBarDisplayMode(selected);
            ApplicationManager.getApplication()
                    .getMessageBus()
                    .syncPublisher(QuotaSettingsListener.TOPIC)
                    .onSettingsChanged();
        }
    }

    @Override
    public void reset() {
        if (displayModeComboBox != null) {
            displayModeComboBox.setSelectedItem(QuotaSettingsState.getInstance().getStatusBarDisplayMode());
        }
        updateAuthUi();
        updateAccountFields();
        updateResponseArea();
    }

    @Override
    public void disposeUIResources() {
        if (connection != null) {
            connection.disconnect();
            connection = null;
        }
        panel = null;
        accountIdField = null;
        emailField = null;
        responseArea = null;
        loginHeaderLabel = null;
        statusLabel = null;
        displayModeComboBox = null;
        loginButton = null;
        cancelLoginButton = null;
        logoutButton = null;
    }

    private void updateAuthUi() {
        QuotaAuthService authService = QuotaAuthService.getInstance();
        boolean loggedIn = authService.isLoggedIn();
        boolean inProgress = authService.isLoginInProgress();
        String statusText = loggedIn ? "Logged in" : "Not logged in";
        if (loginHeaderLabel != null) {
            loginHeaderLabel.setText("Login (" + statusText + ")");
        }
        if (statusLabel != null) {
            statusLabel.setText(statusText);
        }
        loginButton.setEnabled(!inProgress && !loggedIn);
        cancelLoginButton.setEnabled(inProgress);
        logoutButton.setEnabled(loggedIn);
    }

    private void updateResponseArea() {
        if (responseArea == null) {
            return;
        }
        String json = QuotaUsageService.getInstance().getLastResponseJson();
        if (json == null || json.isBlank()) {
            responseArea.setText("No quota response yet.");
        } else {
            responseArea.setText(json);
        }
        responseArea.setCaretPosition(0);
    }

    private void updateAccountFields() {
        if (accountIdField == null || emailField == null) {
            return;
        }
        QuotaAuthService authService = QuotaAuthService.getInstance();
        String accountId = authService.getAccountId();
        String email = null;
        if (authService.isLoggedIn()) {
            de.moritzf.quota.OpenAiCodexQuota quota = QuotaUsageService.getInstance().getLastQuota();
            if (quota != null) {
                email = quota.getEmail();
            }
        }
        accountIdField.setText(accountId == null ? "" : accountId);
        emailField.setText(email == null ? "" : email);
    }
}

File: src/main/kotlin/de/moritzf/quota/idea/QuotaSettingsConfigurable.kt
package de.moritzf.quota.idea

import com.intellij.ide.ActivityTracker
import com.intellij.lang.Language
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ModalityState
import com.intellij.openapi.fileTypes.PlainTextFileType
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.DialogPanel
import com.intellij.openapi.ui.Messages
import com.intellij.ui.EditorTextField
import com.intellij.ui.LanguageTextField
import com.intellij.icons.AllIcons
import com.intellij.ui.components.*
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.RightGap
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.JBColor
import com.intellij.util.messages.MessageBusConnection
import com.intellij.util.ui.JBUI
import com.intellij.util.ui.components.BorderLayoutPanel
import de.moritzf.quota.OpenAiCodexQuota
import de.moritzf.quota.OpenCodeQuota
import javax.swing.Icon
import javax.swing.JButton
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.JTabbedPane
import javax.swing.SwingConstants
import javax.swing.UIManager
import java.awt.Color
import java.awt.Dimension
import java.awt.Toolkit
import java.awt.datatransfer.StringSelection

/**
 * Settings UI that manages authentication actions and shows latest quota payload data.
 */
class QuotaSettingsConfigurable : Configurable {
    private var rootComponent: JComponent? = null
    private var panel: DialogPanel? = null
    private var accountIdField: JBTextField? = null
    private var emailField: JBTextField? = null
    private var openCodeCookieField: JBPasswordField? = null
    private var openCodeStatusLabel: JBLabel? = null
    private var codexResponseViewer: EditorTextField? = null
    private var openCodeResponseViewer: EditorTextField? = null
    private var responseTabbedPane: JTabbedPane? = null
    private var responseViewer: EditorTextField? = null
    private var loginHeaderLabel: JBLabel? = null
    private var statusLabel: JBLabel? = null
    private var statusLabelDefaultForeground: Color? = null
    private var authStatusMessage: AuthStatusMessage? = null
    private var locationComboBox: ComboBox<QuotaIndicatorLocation>? = null
    private var displayModeComboBox: ComboBox<QuotaDisplayMode>? = null
    private var displayModePreview: DisplayModePreviewComponent? = null
    private var loginButton: ActionLink? = null
    private var cancelLoginButton: ActionLink? = null
    private var logoutButton: ActionLink? = null
    private var copyUrlButton: JButton? = null
    private var authUrl: String? = null
    private var connection: MessageBusConnection? = null
    private var updatingDisplayModeChoices: Boolean = false

    override fun getDisplayName(): String = "OpenAI Usage Quota"

    override fun createComponent(): JComponent? {
        accountIdField = JBTextField().apply { isEditable = false }
        emailField = JBTextField().apply { isEditable = false }
        openCodeCookieField = JBPasswordField().apply {
            columns = 40
            toolTipText = "Session cookie from opencode.ai (extract from browser DevTools)"
        }
        openCodeStatusLabel = JBLabel().apply { isVisible = false }
        codexResponseViewer = createResponseViewer()
        openCodeResponseViewer = createResponseViewer()
        responseViewer = codexResponseViewer
        responseTabbedPane = JTabbedPane().apply {
            addTab("OpenAI Codex", codexResponseViewer)
            addTab("OpenCode Go", openCodeResponseViewer)
        }
        loginHeaderLabel = JBLabel()
        statusLabel = JBLabel().apply { isVisible = false }
        statusLabelDefaultForeground = statusLabel!!.foreground ?: UIManager.getColor("Label.foreground")
        locationComboBox = ComboBox(QuotaIndicatorLocation.entries.toTypedArray())
        displayModeComboBox = ComboBox(QuotaDisplayMode.entries.toTypedArray())
        displayModePreview = DisplayModePreviewComponent()
        loginButton = createActionLink("Log In")
        cancelLoginButton = createActionLink("Cancel Login")
        logoutButton = createActionLink("Log Out")
        copyUrlButton = JButton("Copy URL", AllIcons.Actions.Copy).apply {
            isVisible = false
            toolTipText = "Copy login URL to clipboard"
            addActionListener {
                val url = authUrl
                if (!url.isNullOrBlank()) {
                    val selection = StringSelection(url)
                    Toolkit.getDefaultToolkit().systemClipboard.setContents(selection, null)
                }
            }
        }

        locationComboBox!!.addActionListener {
            updateDisplayModeChoices()
            updateDisplayModePreview()
        }

        displayModeComboBox!!.addActionListener {
            if (updatingDisplayModeChoices) {
                return@addActionListener
            }
            updateDisplayModePreview()
        }

        loginButton!!.addActionListener {
            val authService = QuotaAuthService.getInstance()
            if (authService.isLoggedIn()) {
                updateAuthUi()
                return@addActionListener
            }

            loginButton!!.isEnabled = false
            setStatusMessage("Opening browser...")
            authService.startLoginFlow(callback = { result ->
                val currentPanel = panel ?: return@startLoginFlow
                ApplicationManager.getApplication().invokeLater({
                    if (panel == null || statusLabel == null || loginButton == null || logoutButton == null) {
                        return@invokeLater
                    }

                    if (result.success) {
                        setStatusMessage("Logged in")
                        QuotaUsageService.getInstance().refreshNowAsync()
                    } else {
                        val message = result.message ?: "Login failed"
                        val benignFailure = message == "Login canceled" || message == "Logged out"
                        setStatusMessage(if (benignFailure) message else "Login failed", isError = !benignFailure)
                        if (!benignFailure) {
                            Messages.showErrorDialog(currentPanel, message, "OpenAI Login")
                        }
                    }
                    loginButton!!.isEnabled = true
                    updateAuthUi()
                    updateAccountFields()
                }, ModalityState.stateForComponent(currentPanel))
            }, onAuthUrl = { url ->
                val currentPanel = panel ?: return@startLoginFlow
                ApplicationManager.getApplication().invokeLater({
                    if (copyUrlButton == null) return@invokeLater
                    authUrl = url
                    copyUrlButton!!.isVisible = true
                }, ModalityState.stateForComponent(currentPanel))
            })
            updateAuthUi()
        }

        cancelLoginButton!!.addActionListener {
            val aborted = QuotaAuthService.getInstance().abortLogin("Login canceled")
            setStatusMessage(if (aborted) "Login canceled" else "No login in progress")
            updateAuthUi()
        }

        logoutButton!!.addActionListener {
            QuotaAuthService.getInstance().clearCredentials()
            QuotaUsageService.getInstance().clearUsageData("Not logged in")
            setStatusMessage("Logged out")
            updateAuthUi()
            updateAccountFields()
        }

        panel = panel {
            row("Indicator location:") {
                cell(locationComboBox!!)
            }

            row("Indicator display:") {
                cell(displayModeComboBox!!)
                cell(displayModePreview!!).gap(RightGap.SMALL)
            }

            separator()

            row {
                cell(loginHeaderLabel!!)
            }
            row {
                cell(loginButton!!).gap(RightGap.SMALL)
                cell(cancelLoginButton!!).gap(RightGap.SMALL)
                cell(logoutButton!!)
            }
            row {
                cell(statusLabel!!).gap(RightGap.SMALL)
                cell(copyUrlButton!!)
            }

            separator()

            row("Account ID:") {
                cell(accountIdField!!)
                    .resizableColumn()
                    .align(AlignX.FILL)
            }
            row("Email:") {
                cell(emailField!!)
                    .resizableColumn()
                    .align(AlignX.FILL)
            }

            separator()

            row {
                label("OpenCode Go").bold()
            }
            row("Session cookie:") {
                cell(openCodeCookieField!!)
                    .resizableColumn()
                    .align(AlignX.FILL)
                button("Save") {
                    val cookie = String(openCodeCookieField!!.password)
                    if (cookie.isNotBlank()) {
                        OpenCodeSessionCookieStore.getInstance().save(cookie)
                        openCodeCookieField!!.text = ""
                        updateOpenCodeStatus()
                        QuotaUsageService.getInstance().refreshNowAsync()
                    }
                }
                button("Clear") {
                    OpenCodeSessionCookieStore.getInstance().clear()
                    openCodeCookieField!!.text = ""
                    updateOpenCodeStatus()
                    QuotaUsageService.getInstance().clearOpenCodeUsageData()
                }
            }
            row {
                cell(openCodeStatusLabel!!)
            }
            row {
                label("Extract from opencode.ai → DevTools → Storage → Cookies → \"auth\" cookie value. Valid for 1 year.")
            }

            separator()

            onApply {
                val selectedLocation = locationComboBox?.selectedItem as? QuotaIndicatorLocation ?: return@onApply
                val selectedDisplayMode = displayModeComboBox?.selectedItem as? QuotaDisplayMode ?: return@onApply
                val sanitizedDisplayMode = QuotaDisplayMode.sanitizeFor(selectedLocation, selectedDisplayMode)
                val state = QuotaSettingsState.getInstance()
                val locationChanged = selectedLocation != state.location()
                val displayModeChanged = sanitizedDisplayMode != state.displayMode()
                if (locationChanged) {
                    state.setLocation(selectedLocation)
                }
                if (displayModeChanged) {
                    state.setDisplayMode(sanitizedDisplayMode)
                }
                if (locationChanged || displayModeChanged) {
                    ApplicationManager.getApplication().messageBus
                        .syncPublisher(QuotaSettingsListener.TOPIC)
                        .onSettingsChanged()
                    ActivityTracker.getInstance().inc()
                }
            }

            onReset {
                authStatusMessage = null
                locationComboBox?.selectedItem = QuotaSettingsState.getInstance().location()
                updateDisplayModeChoices(QuotaSettingsState.getInstance().displayMode())
                updateDisplayModePreview()
                updateAuthUi()
                updateAccountFields()
                updateResponseArea()
                updateOpenCodeResponseArea()
                updateOpenCodeFields()
            }

            onIsModified {
                val selectedLocation = locationComboBox?.selectedItem as? QuotaIndicatorLocation ?: return@onIsModified false
                val selectedDisplayMode = displayModeComboBox?.selectedItem as? QuotaDisplayMode ?: return@onIsModified false
                val state = QuotaSettingsState.getInstance()
                selectedLocation != state.location() ||
                    QuotaDisplayMode.sanitizeFor(selectedLocation, selectedDisplayMode) != state.displayMode()
            }
        }.apply {
            preferredFocusedComponent = locationComboBox
        }

        rootComponent = BorderLayoutPanel().apply {
            isOpaque = false
            addToTop(panel!!)
            addToCenter(
                BorderLayoutPanel().apply {
                    isOpaque = false
                    border = JBUI.Borders.emptyTop(8)
                    addToTop(
                        JBLabel("Last quota response (JSON):").apply {
                            border = JBUI.Borders.emptyBottom(6)
                        },
                    )
                    addToCenter(responseTabbedPane!!)
                },
            )
        }

        connection = ApplicationManager.getApplication().messageBus.connect()
        connection!!.subscribe(QuotaUsageListener.TOPIC, object : QuotaUsageListener {
            override fun onQuotaUpdated(quota: OpenAiCodexQuota?, error: String?) {
                val currentPanel = rootComponent ?: panel ?: return@onQuotaUpdated
                ApplicationManager.getApplication().invokeLater({
                    if ((rootComponent == null && panel == null) || codexResponseViewer == null || accountIdField == null || emailField == null) {
                        return@invokeLater
                    }
                    updateAccountFields()
                    updateResponseArea()
                    updateOpenCodeStatus()
                }, ModalityState.stateForComponent(currentPanel))
            }

            override fun onOpenCodeQuotaUpdated(quota: OpenCodeQuota?, error: String?) {
                val currentPanel = rootComponent ?: panel ?: return@onOpenCodeQuotaUpdated
                ApplicationManager.getApplication().invokeLater({
                    if (rootComponent == null && panel == null) {
                        return@invokeLater
                    }
                    updateOpenCodeResponseArea()
                    updateOpenCodeStatus()
                }, ModalityState.stateForComponent(currentPanel))
            }
        })

        reset()
        return rootComponent
    }

    override fun isModified(): Boolean {
        return panel?.isModified() == true
    }

    override fun apply() {
        panel?.apply()
    }

    override fun reset() {
        panel?.reset()
    }

    override fun disposeUIResources() {
        connection?.disconnect()
        connection = null
        rootComponent = null
        panel = null
        accountIdField = null
        emailField = null
        openCodeCookieField = null
        openCodeStatusLabel = null
        codexResponseViewer = null
        openCodeResponseViewer = null
        responseTabbedPane = null
        responseViewer = null
        loginHeaderLabel = null
        statusLabel = null
        statusLabelDefaultForeground = null
        authStatusMessage = null
        locationComboBox = null
        displayModeComboBox = null
        displayModePreview = null
        loginButton = null
        cancelLoginButton = null
        logoutButton = null
        copyUrlButton = null
        authUrl = null
        updatingDisplayModeChoices = false
    }

    private fun updateDisplayModeChoices(preferredMode: QuotaDisplayMode? = null) {
        val combo = displayModeComboBox ?: return
        val location = locationComboBox?.selectedItem as? QuotaIndicatorLocation ?: return
        val selectedMode = preferredMode ?: combo.selectedItem as? QuotaDisplayMode ?: QuotaSettingsState.getInstance().displayMode()
        val sanitizedMode = QuotaDisplayMode.sanitizeFor(location, selectedMode)
        updatingDisplayModeChoices = true
        try {
            combo.removeAllItems()
            QuotaDisplayMode.supportedFor(location).forEach(combo::addItem)
            combo.selectedItem = sanitizedMode
        } finally {
            updatingDisplayModeChoices = false
        }
    }

    private fun updateAuthUi() {
        val authService = QuotaAuthService.getInstance()
        val loggedIn = authService.isLoggedIn()
        val inProgress = authService.isLoginInProgress()
        val uiState = QuotaSettingsAuthUiState.create(loggedIn, inProgress, authStatusMessage)
        loginHeaderLabel?.text = uiState.headerText
        loginButton?.isEnabled = uiState.loginEnabled
        cancelLoginButton?.isEnabled = uiState.cancelEnabled
        logoutButton?.isEnabled = uiState.logoutEnabled
        renderStatusMessage(uiState.visibleStatusMessage)
        if (!inProgress) {
            copyUrlButton?.isVisible = false
            authUrl = null
        }
    }

    private fun setStatusMessage(text: String, isError: Boolean = false) {
        authStatusMessage = AuthStatusMessage(text, isError)
    }

    private fun renderStatusMessage(message: AuthStatusMessage?) {
        val label = statusLabel ?: return
        label.text = message?.text.orEmpty()
        label.foreground = if (message?.isError == true) JBColor.RED else statusLabelDefaultForeground ?: label.foreground
        label.isVisible = !message?.text.isNullOrBlank()
    }

    private fun updateResponseArea() {
        val area = codexResponseViewer ?: return
        val json = QuotaUsageService.getInstance().getLastResponseJson()
        area.text = if (json.isNullOrBlank()) "No quota response yet." else json
        area.setCaretPosition(0)
    }

    private fun updateOpenCodeResponseArea() {
        val area = openCodeResponseViewer ?: return
        val quota = QuotaUsageService.getInstance().getLastOpenCodeQuota()
        val error = QuotaUsageService.getInstance().getLastOpenCodeError()
        val json = quota?.rawJson
        area.text = when {
            error != null -> "Error: $error"
            json.isNullOrBlank() -> "No OpenCode response yet."
            else -> json
        }
        area.setCaretPosition(0)
    }

    @Suppress("UsePropertyAccessSyntax")
    private fun createResponseViewer(): EditorTextField {
        val jsonLanguage = Language.findLanguageByID("JSON")
        val viewer = if (jsonLanguage != null) {
            LanguageTextField(jsonLanguage, null, "", false)
        } else {
            EditorTextField("", null, PlainTextFileType.INSTANCE)
        }

        return viewer.apply {
            setOneLineMode(false)
            isViewer = true
            setFontInheritedFromLAF(false)
            preferredSize = Dimension(1, JBUI.scale(220))
            minimumSize = Dimension(1, JBUI.scale(120))
            addSettingsProvider { editor ->
                editor.settings.apply {
                    isLineNumbersShown = false
                    isFoldingOutlineShown = false
                    isLineMarkerAreaShown = false
                    isIndentGuidesShown = false
                    additionalLinesCount = 0
                    additionalColumnsCount = 0
                    isRightMarginShown = false
                }
                editor.setHorizontalScrollbarVisible(true)
                editor.setVerticalScrollbarVisible(true)
            }
        }
    }

    private fun updateAccountFields() {
        val accountField = accountIdField ?: return
        val emailField = emailField ?: return
        val authService = QuotaAuthService.getInstance()
        val accountId = authService.getAccountId()
        val email = if (authService.isLoggedIn()) QuotaUsageService.getInstance().getLastQuota()?.email else null
        accountField.text = accountId.orEmpty()
        emailField.text = email.orEmpty()
    }

    private fun updateOpenCodeFields() {
        openCodeCookieField?.text = ""
        updateOpenCodeStatus()
    }

    private fun updateOpenCodeStatus() {
        val label = openCodeStatusLabel ?: return
        val cookieStore = OpenCodeSessionCookieStore.getInstance()
        val cookie = cookieStore.load()
        val openCodeQuota = QuotaUsageService.getInstance().getLastOpenCodeQuota()
        val openCodeError = QuotaUsageService.getInstance().getLastOpenCodeError()

        when {
            cookie == null -> {
                label.text = "No session cookie configured"
                label.foreground = JBColor.GRAY
            }
            openCodeError != null -> {
                label.text = "Error: $openCodeError"
                label.foreground = JBColor.RED
            }
            openCodeQuota != null -> {
                label.text = "Connected - Go subscription active"
                label.foreground = JBColor.GREEN.darker()
            }
            else -> {
                label.text = "Session cookie stored securely"
                label.foreground = JBColor.GRAY
            }
        }
        label.isVisible = true
    }

    private fun updateDisplayModePreview() {
        val mode = displayModeComboBox?.selectedItem as? QuotaDisplayMode ?: return
        displayModePreview?.updateMode(mode)
    }

    private fun createActionLink(text: String): ActionLink {
        return ActionLink(text).apply {
            autoHideOnDisable = false
        }
    }

    private class DisplayModePreviewComponent : BorderLayoutPanel() {
        private val previewIconLabel = JBLabel().apply {
            horizontalAlignment = SwingConstants.CENTER
            verticalAlignment = SwingConstants.CENTER
        }
        private val percentagePreview = QuotaPercentageIndicator().apply {
            update("42% • 2h 9m", 0.42, QuotaUsageColors.GREEN)
        }

        init {
            isOpaque = false
            border = JBUI.Borders.emptyLeft(4)
            updateMode(QuotaDisplayMode.ICON_ONLY)
        }

        fun updateMode(mode: QuotaDisplayMode) {
            removeAll()
            when (mode) {
                QuotaDisplayMode.ICON_ONLY -> {
                    previewIconLabel.icon = QuotaIcons.STATUS
                    addToCenter(previewIconLabel)
                }

                QuotaDisplayMode.CAKE_DIAGRAM -> {
                    previewIconLabel.icon = scaledCakeIcon(previewIconLabel)
                    addToCenter(previewIconLabel)
                }

                QuotaDisplayMode.PERCENTAGE_BAR -> {
                    addToCenter(percentagePreview)
                }
            }
            revalidate()
            repaint()
        }

        private fun scaledCakeIcon(component: JComponent): Icon {
            return scaleIconToQuotaStatusSize(QuotaIcons.CAKE_40, component)
        }
    }
}

internal data class AuthStatusMessage(
    val text: String,
    val isError: Boolean = false,
)

internal data class QuotaSettingsAuthUiState(
    val headerText: String,
    val visibleStatusMessage: AuthStatusMessage?,
    val loginEnabled: Boolean,
    val cancelEnabled: Boolean,
    val logoutEnabled: Boolean,
) {
    companion object {
        fun create(loggedIn: Boolean, inProgress: Boolean, statusMessage: AuthStatusMessage?): QuotaSettingsAuthUiState {
            val stateText = when {
                inProgress -> "In progress"
                loggedIn -> "Logged in"
                else -> "Not logged in"
            }
            val visibleStatusMessage = statusMessage ?: if (inProgress) {
                AuthStatusMessage("Complete the login in your browser.")
            } else {
                null
            }
            return QuotaSettingsAuthUiState(
                headerText = "Login ($stateText)",
                visibleStatusMessage = visibleStatusMessage,
                loginEnabled = !inProgress && !loggedIn,
                cancelEnabled = inProgress,
                logoutEnabled = loggedIn,
            )
        }
    }
}
